### PR TITLE
Support COPY TO Externally Defined File Formats, add FileType trait

### LINKED ISSUE #8345: enable users to add support for LIstingTable /  object_store table formats of different types
### Is your feature request related to a problem or challenge?

I've been exploring adding new file formats and types to our system (@GlareDB), and have run into the problem where the `FileFormat` trait depends on the `FileType` enum, which means it's impossible to implement entirely new filetypes except from within the package. 

This is sort of gnarly to unwind as I've dug in, and I don't have a particular solution that I favor. 

### Describe the solution you'd like

I think collapsing `FileType` into `FileFormat` is maybe my favorite solution conceptually, but I think (more narrowly) removing the references to `FileType` in the `FileFormat` (as in my branch) would be good. 

### Describe alternatives you've considered

There are a few ways to resolve this:
- make the FileType an implementable trait rather than an enum.
- remove dependency between the interfaces [I tried to do this here](https://github.com/apache/arrow-datafusion/compare/main...tychoish:arrow-datafusion:file-type-and-format), which I think is a good start, but I think the issue runs even deeper. Having said that, the only casualty here would be the listing code, which is maybe something that can be worked around.
- move all properties of "file_type" into the `FileFormat` (which kind of makes sense? I'm not sure that these are actually meaningfully separate abstractions.)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by devinjdangelo**
## Which issue does this PR close?

Closes #8345
Revives #8667 

## Rationale for this change

Previously, `COPY TO` queries could only be planned/executed for a preset enum of `FileTypes`. `CREATE EXTERNAL TABLE` supports arbitrary file formats via the `FileFormat` and `TableProvider` traits. This PR removes the `FileType` enum in favor of a `FileType` trait and updates `COPY TO` plans to work with any externally implemented `dyn FileFormat`.  

This PR also enables future work to remove csv, parquet, json, ect... formats into dedicated crates (datafusion-csv, ...). 

## What changes are included in this PR?

- Deletes the `FileType` enum and replaces with a trait
- Creates a wrapper struct for `FileFormat` that implements `FileType` to enable conversions between the two
- Extends `LogicalExtensionCodec` to allow serde of `dyn FileType` 

## Are these changes tested?

Existing tests verify behavior for currently supported formats remains the same.

## Are there any user-facing changes?

Yes, there are API changes relevant for developers using DataFusion.

### CODE REVIEW

**Comment by alamb** [comment]
This looks awesome -- thank you @devinjdangelo  -- I plan to review this carefully, but probably not until tomorrow

File: datafusion/expr/src/logical_plan/dml.rs
Comment (devinjdangelo) [inline datafusion/expr/src/logical_plan/dml.rs:39]: If using DataFusion `core` for execution plans, this `FileType` can be converted to a `FileFormatFactory` which will produce a `FileFormat` at runtime.

**Review by devinjdangelo — COMMENTED** [review]
cc @alamb @thinkharderdev I finally had some time to figure out how to make #8667 work. 

There are a number of tasks to be done, preferably as follow on PRs given the size/complexity of the changes in this PR already:

- [ ] end-to-end example of implementing a custom `FileFormat` and using `COPY` with it
- [ ]  Finishing `LogicalExtensionCodec` impls for csv, json, parquet, arrow, and avro `FileFormat`
- [ ] Moving `FileFormat` impls  and supporting code out of core and into dedicated crates
- [ ] (maybe) Moving `ListingTable` to a dedicated crate

File: datafusion/core/src/datasource/file_format/mod.rs
Comment (alamb) [inline datafusion/core/src/datasource/file_format/mod.rs:53]:
```suggestion
/// Factory for creating [`FileFormat`] instances based on session and command level options
///
/// Users can provide their own `FileFormatFactory` to support arbitrary file formats
```

File: datafusion/core/src/datasource/file_format/mod.rs
Comment (alamb) [inline datafusion/core/src/datasource/file_format/mod.rs:78]: Minor: Is there any reason this has to return an owned `String`? Could it return `&str` ?

File: datafusion/core/src/datasource/file_format/mod.rs
Comment (alamb) [inline datafusion/core/src/datasource/file_format/mod.rs:84]: Same question about owned String

File: datafusion/core/src/datasource/file_format/mod.rs
Comment (alamb) [reply datafusion/core/src/datasource/file_format/mod.rs:139]: I found the naming confusing -- I would have expected that `DefaultFileFormat` implemented `FileFormatFactory` or something

Maybe we could call this `FileFormatSource` ?

File: datafusion/core/src/datasource/file_format/mod.rs
Comment (alamb) [inline datafusion/core/src/datasource/file_format/mod.rs:152]: Rather than this wrapper, I wonder if we could implement `FileType` for `dyn FileFormatFactory`

```rust
impl FileType for &dyn FileFormatFactory{ 
...
}
```

And avoid having to have this struct 🤔

File: datafusion/core/src/execution/session_state.rs
Comment (alamb) [inline datafusion/core/src/execution/session_state.rs:251]: 👍

File: datafusion/core/src/execution/session_state.rs
Comment (alamb) [inline datafusion/core/src/execution/session_state.rs:256]: Another way to write this would be

```suggestion
        if let Err(e) = new_self.register_file_format(Arc::new(ParquetFormatFactory::new()), false)
        {
           log::info!("Unable to register default ParquetFormat: {e}"),
        };
```

which might be more concise 🤔

File: datafusion/core/src/physical_planner.rs
Comment (alamb) [inline datafusion/core/src/physical_planner.rs:789]: this is great this code has been removed from the default planner

**Review by alamb — COMMENTED** [review]
Thank you @devinjdangelo -- I had a good look at this PR, and I think it looks really nice

I did not quite get through the entire PR yet, but I wanted to provide some early feeback

I think it very much adheres to DataFusion's design goals of permiting extension at all points and makes it easier to support new file formats

One thing I didn't fully grok at first was whyn there was a need for `FileFormatFactory` -- like why not simply use `FileFormat` everywhere.

Then I realized that the reason was that the `FileFormat`s have state, specifically default options and that the format itself then carries the specific options


```rust
pub struct JsonFormatFactory {
    /// default options
    options: Option<JsonOptions>,  
}
```

Perhaps we can document this rationale somewhere so others are not confused as well.


## Suggestion: Add an example
I think something that would be awesome to add (as a follow on PR perhaps) is an example of creating a custom file format and read/write to it. 

https://github.com/apache/datafusion/issues/11079

File: datafusion/common/src/config.rs
Comment (alamb) [reply datafusion/common/src/config.rs:1122]: 🤔  perhaps we should add some Trait to unify the handling of options for built in formats and custom formats 🤔

File: datafusion/common/src/config.rs
Comment (alamb) [inline datafusion/common/src/config.rs:1690]: it is certainly nice to avoid many of these `            #[cfg(feature = "parquet")]` that were sprinkled throughout the code

File: datafusion/common/src/config.rs
Comment (alamb) [inline datafusion/common/src/config.rs:1254]: the parameter is called "format" but the struct is `ConfigFileType`

Maybe we should call this `set_config_type`? Or maybe it could just pass in a `Arc<dyn FIleFormat>~?

File: datafusion/core/src/datasource/file_format/file_compression_type.rs
Comment (alamb) [inline datafusion/core/src/datasource/file_format/file_compression_type.rs:286]: Should we port these tests instead of just removing them?

File: datafusion/expr/src/logical_plan/dml.rs
Comment (alamb) [reply datafusion/expr/src/logical_plan/dml.rs:39]: What is the reason it can't be `FileFormatFactory` directly 🤔

File: datafusion/proto/src/logical_plan/file_formats.rs
Comment (alamb) [inline datafusion/proto/src/logical_plan/file_formats.rs:108]: do we have to complete the TODO?

**Review by alamb — COMMENTED** [review]
Thank you so much @devinjdangelo  -- I made it through this PR and I think it looks really nice

I do think it is worth considering how to simplify the traits (possibly unifying the configuration) and there look like there are a few todos, but otherwise this PR looks really nice to me

cc @ozankabak @berkaysynnada @metesynnada

File: datafusion/core/src/datasource/file_format/file_compression_type.rs
Comment (devinjdangelo) [reply datafusion/core/src/datasource/file_format/file_compression_type.rs:286]: Yes, I moved this functionality to a trait method on each `FileFormatFactory`. We could add a test for each format which is similar to this test.

File: datafusion/core/src/datasource/file_format/mod.rs
Comment (devinjdangelo) [reply datafusion/core/src/datasource/file_format/mod.rs:80]: I think &str would be too restrictive, but we could avoid requiring a String allocation with a cow instead.

File: datafusion/core/src/datasource/file_format/mod.rs
Comment (devinjdangelo) [reply datafusion/core/src/datasource/file_format/mod.rs]: I renamed this to `DefaultFileType`.

File: datafusion/core/src/datasource/file_format/mod.rs
Comment (devinjdangelo) [reply datafusion/core/src/datasource/file_format/mod.rs]: I tried briefly, but ran into various errors. I am not sure if it is possible with the traits defined in different crates to make this work without a wrapper. I think that is why `DefaultTableSource` exists as well.

File: datafusion/expr/src/logical_plan/dml.rs
Comment (devinjdangelo) [reply datafusion/expr/src/logical_plan/dml.rs:39]: This would introduce a dependency between datafusion-expr and datafusion-core. The `FileType` trait exists to avoid this dependency and so users only using datafusion for logical planning would not have to implement all of the physical execution related `FileFormat`methods.

File: datafusion/proto/src/logical_plan/file_formats.rs
Comment (devinjdangelo) [reply datafusion/proto/src/logical_plan/file_formats.rs:108]: Users depending on the ability to serialize COPY plans (e.g. ballista) will need this TODO to be completed before upgrading to any version of datafusion including this change.

I think it would be OK to cut an follow up ticket for this.

File: datafusion/common/src/config.rs
Comment (devinjdangelo) [reply datafusion/common/src/config.rs:1122]: We could potentially remove the `TableOptions` code and instead have each `FileFormatFactory` handle configuration. This is actually mostly the case in this PR already, but `TableOptions` is a common helper.

**Review by devinjdangelo — COMMENTED** [review]
Thank you @alamb for the thorough review as always! I have made a few small changes.

**Comment by berkaysynnada** [comment]
Thank you, @devinjdangelo, for proposing this extensibility feature. I would like to clarify a few points to ensure I understand correctly:

* We cannot merge `FileType` and `FileFormatFactory` due to dependency issues.
* We cannot merge `FileFormatFactory` and `FileFormat` because of stateful format factories.

If these points are correct, I am considering whether we could move these states (extension options, like `CsvOptions`) into `FileType` somehow to remove one of these layers. This is just a thought experiment; I don't want to block the PR, which I believe improves extensibility.

**Comment by devinjdangelo** [comment]
Thank you for taking a look @berkaysynnada 

> We cannot merge FileType and FileFormatFactory due to dependency issues.

That is correct. Logical planning cannot depend on FileFormatFactory or FileFormat to avoid a dependency on datafusion's execution engine during logical planning.

> We cannot merge FileFormatFactory and FileFormat because of stateful format factories.

Yes, `SessionState` stores a `HashMap<String, Arc<dyn FileFormatFactory>`. Each `FileFormatFactory` could be shared among multiple concurrent queries and must be able to produce many, differently configured, `Arc<dyn FileFormat>`s. If `SessionState` stored only a `Arc<dyn FileFormat>` it would have to have a `&mut self` method enabled by a `Mutex<>` and even with that it would possibly lead to unexpected behavior if multiple queries modified the format during execution.

**Comment by alamb** [comment]
> Yes, SessionState stores a HashMap<String, Arc<dyn FileFormatFactory>>. Each FileFormatFactory could be shared among multiple concurrent queries and must be able to produce many, differently configured, 

I wonder if it wuld be possible to store the file format state outside the `FileFormat` and pass in the format options as part of each query somehow 🤔

**Review by alamb — COMMENTED** [review]
Lets see if we can find some way to avoid the extra level of stateful factories. If not then I think this PR could still be merged

Before we merge I'll file follow on tickets for stateful protobuf encoding

**Comment by devinjdangelo** [comment]
> I wonder if it wuld be possible to store the file format state outside the FileFormat and pass in the format options as part of each query somehow 🤔

I think this is possible, but it would also make implementing `FileFormat` a bit inconvenient. We would have to pass in and parse `format_options` on every trait method. E.g. something like:

```rust
pub trait FileFormat{
        async fn create_physical_plan(
        &self,
        state: &SessionState,
        conf: FileScanConfig,
        filters: Option<&Arc<dyn PhysicalExpr>>,
        format_options: &HashMap<String, String>,
    ) -> Result<Arc<dyn ExecutionPlan>>;
    ...
}
```

And each method would have to parse the `format_options`. `FileFormatFactory` avoids this by allowing `FileFormat` to be created during physical planning when query specific options are known and can be parsed once/stored in the `FileFormat` struct.

**Review by alamb — APPROVED** [review]
I think we should proceed with this PR, even though some of the traits are less than ideal (due to having to handle state)

Also before merging we should file the follow on tickets you suggest in https://github.com/apache/datafusion/pull/11060#pullrequestreview-2133785636

Let's leave it open for another day or two to gather additional comments

Thanks again for pushing this part of the code forward

File: datafusion/proto/src/logical_plan/file_formats.rs
Comment (alamb) [reply datafusion/proto/src/logical_plan/file_formats.rs:34]: I will file follow on tickets

**Comment by alamb** [comment]
I merged up to resolve a conflict

**Review by alamb — APPROVED** [review]
Filed https://github.com/apache/datafusion/issues/11150 to track moving the extension codes

I plan to merge this PR when CI is complete

**Comment by alamb** [comment]
Forward ! Thank you @devinjdangelo  and @Weijun-H

**Pull request merged**
